cpp/include/cudf/dfa/superstate.hpp

```diff
@@ -0,0 +1,226 @@
+ * // alias superstate for convienence.
+ * // csv_state has 7 unique states, and will therefore require a total of 4 bytes (uint32_t)
+ * using csv_superstate = superstate<csv_state, csv_token, 7>;
+ *
+ * // apply state transitions in "parallel"
+ * auto a = csv_superstate() + csv_token::comment + csv_token::other + csv_token::newline;
+ * auto b = csv_superstate() + csv_token::other + csv_token::delimiter;
+ * auto c = csv_superstate() + csv_token::other + csv_token::delimiter;
+ *
+ * // given record_end as the initial state, get the resulting state after all transitions.
+ * auto final = a + b + c;
+ * auto result_a = final.get(csv_state::record_end);   // csv_state::field_end
+ * auto result_b = final.get(csv_state::quoted_field); // csv_state::quoted_field
+ * auto result_c = final.get(csv_state::field_end);    // csv_state::invalid
+ * ```
+ *
+ * @note this class is constexpr, and can be used on both host and device
+ *
+ * @tparam State state which will be tracked - up to 16 unique values. usually an enum class.
+ * @tparam Instruction type which can be added to a state to transition in to a new state.
+ * @tparam N total number of states representable by State, up to 16 unique states.
+ */
+template <typename State, typename Instruction, uint8_t N>
+struct superstate {
+ private:
+  static const uint8_t BITS = superstate_policy<N>::BITS;
+  static const uint8_t MASK = superstate_policy<N>::MASK;
+
+  using Data = typename superstate_policy<N>::Data;
+
+  Data _data;
+
+ public:
+  /**
+   * @brief creates a superstate which represents all possible states and applied transitions
+   */
+  inline constexpr superstate() : _data(0)
+  {
+    for (auto i = 0; i < N; i++) { _data |= i << (i * BITS); }
+  }
+
+  /**
+   * @brief creates a superstate which represents only a single state
+   */
+  inline constexpr superstate(State state) : _data(0)
+  {
+    for (auto i = 0; i < N; i++) { _data |= static_cast<Data>(state) << (i * BITS); }
+  }
+
+  /**
+   * @brief create a superstate with a specific underlying value
+   * @param data the underlying value
+   */
+  inline constexpr superstate(Data data) : _data(data) {}
+
+  inline constexpr Data data() const {
+    return _data;
+  }
+
+  /**
+   * @brief converts to a single state using the least significant bits of the underlying data
+   *
+   * @return State a state created from the least significant bits of the underlying data
+   */
+  inline constexpr operator State() const { return static_cast<State>(_data & MASK); }
+
+  /**
+   * @brief transitions each underlying state using the provided instruction
+   *
+   * @note the + operator is called N times, once for each state.
+   *
+   * @param rhs the instruction used to transition all states
+   * @return the result of all transitions
+   */
+  inline constexpr superstate operator+(Instruction rhs) const
+  {
+    superstate result(0);
+    for (auto i = 0; i < N; i++) { result.set<false>(i, get(i) + rhs); }
+    return result;
+  }
```

Comment:
I considered removing the `Instruction` typeparam from the struct and moving here, but that seemed overly generic. Another option I considered was removing the `Instruction` typeparam altogether, converting this operator to a named member function and, and having it accept a lambda (`(State) -> State`) that would be applied to each state.

Reply:
I'd remove the `Instruction` type param from the class and make the `operator+` a free function instead of a member function.